### PR TITLE
Throw NoFallbackError instead of returning

### PR DESCRIPTION
**Pull request by devknoll**
The `false` variation can't cleanly be modeled in a `(req, res) => Promise<void>` universe, so lets change it to throw an error for now. We can probably remove this in the future.

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| buildDuration | 10.9s | 10.9s | -6ms |
| nodeModulesSize | 56.3 MB | 56.3 MB | ⚠️ +200 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..67cb.js gzip | 9.69 kB | 9.69 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 55.3 kB | 55.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.79 kB | 4.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.64 kB | 6.64 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.3 kB | 51.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 18.9 kB | 18.9 kB | ✓ |
| Overall change | 18.9 kB | 18.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _app.js gzip | 1.07 kB | 1.07 kB | ✓ |
| _error.js gzip | 2.98 kB | 2.98 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 1.89 kB | 1.89 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.38 kB | 7.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _app.module.js gzip | 589 B | 589 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.27 kB | 5.27 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| index.html gzip | 936 B | 936 B | ✓ |
| link.html gzip | 946 B | 946 B | ✓ |
| withRouter.html gzip | 934 B | 934 B | ✓ |
| Overall change | 2.82 kB | 2.82 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| buildDuration | 11.6s | 11.6s | -55ms |
| nodeModulesSize | 56.3 MB | 56.3 MB | ⚠️ +200 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..67cb.js gzip | 9.69 kB | 9.69 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 55.3 kB | 55.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.79 kB | 4.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.64 kB | 6.64 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.3 kB | 51.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 18.9 kB | 18.9 kB | ✓ |
| Overall change | 18.9 kB | 18.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _app.js gzip | 1.07 kB | 1.07 kB | ✓ |
| _error.js gzip | 2.98 kB | 2.98 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 1.89 kB | 1.89 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.38 kB | 7.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _app.module.js gzip | 589 B | 589 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.27 kB | 5.27 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _error.js gzip | 292 kB | 292 kB | -218 B |
| 404.html gzip | 1.34 kB | 1.34 kB | ✓ |
| hooks.html gzip | 975 B | 975 B | ✓ |
| index.js gzip | 292 kB | 292 kB | ⚠️ +533 B |
| link.js gzip | 299 kB | 299 kB | ⚠️ +27 B |
| routerDirect.js gzip | 298 kB | 298 kB | -165 B |
| withRouter.js gzip | 299 kB | 299 kB | -3 B |
| Overall change | 1.48 MB | 1.48 MB | ⚠️ +174 B |

</details>

</details>


Commit: 500db9717aac40958b3b87b9c00948590deb0000

**Review by ijjk — APPROVED** [review]
LGTM, will leave for @Timer to also review

File: packages/next/next-server/server/next-server.ts [hunk 1]
```diff
@@ -1237,10 +1243,6 @@ export default class Server {
           err,
         }
       )
-      if (result2 === false) {
-        throw new Error('invariant: failed to render error page')
```
Comment (Timer) [inline packages/next/next-server/server/next-server.ts]: Please re-introduce this invariant in the catch clause.

**Review by Timer — CHANGES_REQUESTED** [review]
While I'm not a fan of using errors as control-flow (and TypeScript reasons), I understand the [temporary] necessity of this change.

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| buildDuration | 10.7s | 10.6s | -139ms |
| nodeModulesSize | 56.3 MB | 56.3 MB | ⚠️ +658 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..67cb.js gzip | 9.69 kB | 9.69 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 55.3 kB | 55.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.79 kB | 4.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.64 kB | 6.64 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.3 kB | 51.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 18.9 kB | 18.9 kB | ✓ |
| Overall change | 18.9 kB | 18.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _app.js gzip | 1.07 kB | 1.07 kB | ✓ |
| _error.js gzip | 2.98 kB | 2.98 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 1.89 kB | 1.89 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.38 kB | 7.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _app.module.js gzip | 589 B | 589 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.27 kB | 5.27 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| index.html gzip | 936 B | 936 B | ✓ |
| link.html gzip | 946 B | 946 B | ✓ |
| withRouter.html gzip | 934 B | 934 B | ✓ |
| Overall change | 2.82 kB | 2.82 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| buildDuration | 11.3s | 11.5s | ⚠️ +145ms |
| nodeModulesSize | 56.3 MB | 56.3 MB | ⚠️ +658 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..67cb.js gzip | 9.69 kB | 9.69 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 55.3 kB | 55.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.79 kB | 4.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.64 kB | 6.64 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.3 kB | 51.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 18.9 kB | 18.9 kB | ✓ |
| Overall change | 18.9 kB | 18.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _app.js gzip | 1.07 kB | 1.07 kB | ✓ |
| _error.js gzip | 2.98 kB | 2.98 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 1.89 kB | 1.89 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.38 kB | 7.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _app.module.js gzip | 589 B | 589 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.27 kB | 5.27 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | devknoll/next.js add-next-request-response | Change |
| - | - | - | - |
| _error.js gzip | 292 kB | 292 kB | -534 B |
| 404.html gzip | 1.34 kB | 1.34 kB | ✓ |
| hooks.html gzip | 975 B | 975 B | ✓ |
| index.js gzip | 292 kB | 292 kB | -215 B |
| link.js gzip | 299 kB | 300 kB | ⚠️ +888 B |
| routerDirect.js gzip | 299 kB | 299 kB | ⚠️ +43 B |
| withRouter.js gzip | 298 kB | 298 kB | -173 B |
| Overall change | 1.48 MB | 1.48 MB | ⚠️ +9 B |

</details>

</details>

**Review by Timer — APPROVED** [review]

**Pull request merged**